src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -596,7 +597,9 @@ private bool TakeStartLine(SocketInput input)
                 {
                     return false;
                 }
-                var requestUri = begin.GetString(scan);
+
+                var beginPath = begin;
+                var endPath = scan;
```

Comment:
the line above `scan.Seek(' ', '?')` should be changed to `scan.Seek(' ', '?', '%', '+')` - only if chFound returns '%' or '+' would you set a flag `needsDecoding` and call `scan.Seek(' ', '?')` to find the real end of the path portion. in this way you skip the entire decoding step altogether for the majority of paths that contain entirely face-value bytes